pkg/crd-schema/gen.go

```diff
@@ -0,0 +1,352 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crdschema
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	crdgen "sigs.k8s.io/controller-tools/pkg/crd"
+	crdschemayaml "sigs.k8s.io/controller-tools/pkg/crd-schema/yaml"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	utilruntime.Must(apiextensionsv1beta1.AddToScheme(scheme))
+}
+
+// Generator is a genall.Generator that generates CRD schemas.
+type Generator struct {
+	// ManifestPath is a path with CustomResourceDefinition manifest files
+	ManifestPath string
+	// If true, no file is written. The Generate func will return an
+	// error if a file on disk differs from the generated manifest.
+	VerifyOnly bool
+}
+
+func (Generator) RegisterMarkers(into *markers.Registry) error {
+	return crdmarkers.Register(into)
+}
+
+func (g Generator) Generate(ctx *genall.GenerationContext) error {
+	p := &crdgen.Parser{
+		Collector: ctx.Collector,
+		Checker:   ctx.Checker,
+	}
+	f := crdgen.Flattener{
+		Parser: p,
+	}
+
+	crdgen.AddKnownTypes(p)
+	for _, root := range ctx.Roots {
+		p.NeedPackage(root)
+	}
+
+	metav1Pkg := crdgen.FindMetav1(ctx.Roots)
+	if metav1Pkg == nil {
+		// no objects in the roots, since nothing imported metav1
+		fmt.Printf("No metav1 import found in any of the API packages.\n")
+		return nil
+	}
+
+	// load existing manifests from manifests/ dir
+	existingCRDsByFileNames, err := crdsFromDirectory(g.ManifestPath)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Found %d CRD manifests.\n", len(existingCRDsByFileNames))
+	existingFileNamesByGroupKind := map[schema.GroupKind]string{}
+	for fn, crd := range existingCRDsByFileNames {
+		existingFileNamesByGroupKind[crd.GroupKind] = fn
+	}
+
+	changedFiles := []string{}
+	for _, groupKind := range crdgen.FindKubeKinds(p, metav1Pkg) {
+		fn, found := existingFileNamesByGroupKind[groupKind]
+		if !found {
+			continue
+		}
+		existing := existingCRDsByFileNames[fn]
+
+		// get existing versions
+		existingVersions := make(map[string]bool, 0)
+		if existing.CRD.Spec.Version != "" {
+			existingVersions[existing.CRD.Spec.Version] = true
+		}
+		for _, v := range existing.CRD.Spec.Versions {
+			existingVersions[v.Name] = true
+		}
+
+		// go through source code versions
+		versionSchemas := map[string]*apiextensionsv1beta1.JSONSchemaProps{}
+		for pkg, gv := range p.GroupVersions {
```

Comment:
probably clearer if you just construct `gv --> pkg` as a map ahead of time.

Reply:
some of this logic needs to be broken out into helpers as well.

Reply:
not sure I can follow. This is collecting the schemas for the versions of the current CRD. Why gv?